diff --git a/thirdparty/thorvg/src/common/tvgCommon.h b/thirdparty/thorvg/src/common/tvgCommon.h
index 316fb8dc0b..83da2a19ae 100644
--- a/thirdparty/thorvg/src/common/tvgCommon.h
+++ b/thirdparty/thorvg/src/common/tvgCommon.h
@@ -63,9 +63,6 @@ using namespace tvg;
     #define TVG_FALLTHROUGH
 #endif
 
-void* operator new(std::size_t size);
-void operator delete(void* ptr) noexcept;
-
 namespace tvg
 {
     enum class FileType
diff --git a/thirdparty/thorvg/src/renderer/tvgInitializer.cpp b/thirdparty/thorvg/src/renderer/tvgInitializer.cpp
index 17b771102e..6a8ce97f7d 100644
--- a/thirdparty/thorvg/src/renderer/tvgInitializer.cpp
+++ b/thirdparty/thorvg/src/renderer/tvgInitializer.cpp
@@ -131,27 +131,3 @@ uint16_t THORVG_VERSION_NUMBER()
 {
     return _version;
 }
-
-
-void* operator new(std::size_t size)
-{
-    return tvg::malloc(size);
-}
-
-
-void operator delete(void* ptr) noexcept
-{
-    tvg::free(ptr);
-}
-
-
-void* operator new[](std::size_t size)
-{
-    return tvg::malloc(size);
-}
-
-
-void operator delete[](void* ptr) noexcept
-{
-    tvg::free(ptr);
-}
