diff --git a/thirdparty/thorvg/src/loaders/lottie/tvgLottieParser.cpp b/thirdparty/thorvg/src/loaders/lottie/tvgLottieParser.cpp
index a27c4c8065..f402fca3d7 100644
--- a/thirdparty/thorvg/src/loaders/lottie/tvgLottieParser.cpp
+++ b/thirdparty/thorvg/src/loaders/lottie/tvgLottieParser.cpp
@@ -946,9 +946,10 @@ LottieImage* LottieParser::parseImage(const char* data, const char* subPath, boo
         image->size = b64Decode(b64Data, length, &image->b64Data);
     //external image resource
     } else {
-        auto len = strlen(dirName) + strlen(subPath) + strlen(data) + 1;
+        const char* dir = dirName == nullptr ? "" : dirName;
+        auto len = strlen(dir) + strlen(subPath) + strlen(data) + 1;
         image->path = static_cast<char*>(malloc(len));
-        snprintf(image->path, len, "%s%s%s", dirName, subPath, data);
+        snprintf(image->path, len, "%s%s%s", dir, subPath, data);
     }
 
     image->width = width;
