 thirdparty/thorvg/src/renderer/tvgRender.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/thirdparty/thorvg/src/renderer/tvgRender.h b/thirdparty/thorvg/src/renderer/tvgRender.h
index 8f28d37dbc..a915d68fec 100644
--- a/thirdparty/thorvg/src/renderer/tvgRender.h
+++ b/thirdparty/thorvg/src/renderer/tvgRender.h
@@ -103,7 +103,7 @@ struct RenderRegion
     void intersect(const RenderRegion& rhs);
     void add(const RenderRegion& rhs);
 
-    bool operator==(const RenderRegion& rhs)
+    bool operator==(const RenderRegion& rhs) const
     {
         if (x == rhs.x && y == rhs.y && w == rhs.w && h == rhs.h) return true;
         return false;
