diff --git a/thirdparty/sdl/stdlib/SDL_getenv.c b/thirdparty/sdl/stdlib/SDL_getenv.c
index 17220c69ae..50a511262b 100644
--- a/thirdparty/sdl/stdlib/SDL_getenv.c
+++ b/thirdparty/sdl/stdlib/SDL_getenv.c
@@ -555,7 +555,7 @@ int SDL_unsetenv_unsafe(const char *name)
 #endif // HAVE_LIBC_ENVIRONMENT
 
 // Retrieve a variable named "name" from the environment
-#ifdef HAVE_LIBC_ENVIRONMENT
+#if defined(SDL_PLATFORM_LINUX) || defined(SDL_PLATFORM_MACOS) || defined(SDL_PLATFORM_FREEBSD)
 const char *SDL_getenv_unsafe(const char *name)
 {
 #ifdef SDL_PLATFORM_ANDROID
