diff --git a/thirdparty/sdl/joystick/apple/SDL_mfijoystick.m b/thirdparty/sdl/joystick/apple/SDL_mfijoystick.m
index c7ed381d66..0956a803f8 100644
--- a/thirdparty/sdl/joystick/apple/SDL_mfijoystick.m
+++ b/thirdparty/sdl/joystick/apple/SDL_mfijoystick.m
@@ -326,14 +326,14 @@ static bool IOS_AddMFIJoystickDevice(SDL_JoystickDeviceItem *device, GCControlle
      * struct, and ARC doesn't work with structs. */
     device->controller = (__bridge GCController *)CFBridgingRetain(controller);
 
-    if (controller.vendorName) {
-        name = controller.vendorName.UTF8String;
-    } else {
-        if (@available(macOS 10.15, iOS 13.0, tvOS 13.0, *)) {
+    if (@available(macOS 10.15, iOS 13.0, tvOS 13.0, *)) {
             if (controller.productCategory) {
                 name = controller.productCategory.UTF8String;
             }
-        }
+        } else {
+            if (controller.vendorName) {
+                name = controller.vendorName.UTF8String;
+            }
     }
 
     if (!name) {
