diff --git a/thirdparty/sdl/joystick/windows/SDL_dinputjoystick.c b/thirdparty/sdl/joystick/windows/SDL_dinputjoystick.c
index 62ffbb46ec..be94c07845 100644
--- a/thirdparty/sdl/joystick/windows/SDL_dinputjoystick.c
+++ b/thirdparty/sdl/joystick/windows/SDL_dinputjoystick.c
@@ -444,17 +444,11 @@ bool SDL_DINPUT_JoystickInit(void)
         // The 64-bit version of EZFRD64.DLL crashes after being loaded,
         // which happens implicitly when querying the device capabilities,
         // so make sure we don't do that if there's a possibility of crashing
-        static const char *directories[] = {
-            "C:/Windows/USB_Vibration",
-            "C:/Windows/USB Vibration"
-        };
-        for (int i = 0; i < SDL_arraysize(directories) && !has_broken_EZFRD64DLL; ++i) {
-            int count = 0;
-            char **files = SDL_GlobDirectory(directories[i], "*/EZFRD64.DLL", SDL_GLOB_CASEINSENSITIVE, &count);
-            if (count > 0) {
-                has_broken_EZFRD64DLL = true;
-            }
-            SDL_free(files);
+        DWORD a1 = GetFileAttributesW(L"C:\\Windows\\USB_Vibration");
+        DWORD a2 = GetFileAttributesW(L"C:\\Windows\\USB Vibration");
+
+        if ((a1 != INVALID_FILE_ATTRIBUTES && (a1 & FILE_ATTRIBUTE_DIRECTORY)) || (a2 != INVALID_FILE_ATTRIBUTES && (a2 & FILE_ATTRIBUTE_DIRECTORY))) {
+            has_broken_EZFRD64DLL = true;
         }
         if (has_broken_EZFRD64DLL) {
             SDL_LogWarn(SDL_LOG_CATEGORY_INPUT, "Broken EZFRD64.DLL detected, disabling DirectInput force feedback");
