diff --git a/thirdparty/mbedtls/tf-psa-crypto/include/mbedtls/pk.h b/thirdparty/mbedtls/tf-psa-crypto/include/mbedtls/pk.h
index 3eb4dc2d26..de56242c11 100644
--- a/thirdparty/mbedtls/tf-psa-crypto/include/mbedtls/pk.h
+++ b/thirdparty/mbedtls/tf-psa-crypto/include/mbedtls/pk.h
@@ -88,7 +88,7 @@ typedef struct mbedtls_pk_info_t mbedtls_pk_info_t;
 #define MBEDTLS_PK_MAX_RSA_PUBKEY_RAW_LEN \
     PSA_KEY_EXPORT_RSA_PUBLIC_KEY_MAX_SIZE(PSA_VENDOR_RSA_MAX_KEY_BITS)
 
-#define MBEDTLS_PK_MAX_PUBKEY_RAW_LEN 0
+#define MBEDTLS_PK_MAX_PUBKEY_RAW_LEN 1
 #if defined(PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY) && \
     MBEDTLS_PK_MAX_EC_PUBKEY_RAW_LEN > MBEDTLS_PK_MAX_PUBKEY_RAW_LEN
 #undef MBEDTLS_PK_MAX_PUBKEY_RAW_LEN
