diff --git a/thirdparty/basis_universal/encoder/basisu_enc.h b/thirdparty/basis_universal/encoder/basisu_enc.h
index a565803e04..5256d34ad3 100644
--- a/thirdparty/basis_universal/encoder/basisu_enc.h
+++ b/thirdparty/basis_universal/encoder/basisu_enc.h
@@ -2164,10 +2164,9 @@ namespace basisu
 		uint32_t max_threads, job_pool *pJob_pool,
 		bool even_odd_input_pairs_equal)
 	{
-		typedef bit_hasher<typename Quantizer::training_vec_type> training_vec_bit_hasher;
-
 		// rg 6/24/2025 - Cross platform determinism
 #if 0
+		typedef bit_hasher<typename Quantizer::training_vec_type> training_vec_bit_hasher;
 		typedef std::unordered_map < typename Quantizer::training_vec_type, weighted_block_group, 
 			training_vec_bit_hasher> group_hash;
 #else
