diff --git a/thirdparty/vulkan/vk_enum_string_helper.h b/thirdparty/vulkan/vk_enum_string_helper.h
index 8c77dd3ee5..74f49a6430 100644
--- a/thirdparty/vulkan/vk_enum_string_helper.h
+++ b/thirdparty/vulkan/vk_enum_string_helper.h
@@ -22,8 +22,10 @@
 ****************************************************************************/
 // NOLINTBEGIN
 #pragma once
+#ifdef __cplusplus
 #include <string>
-#include <vulkan/vulkan.h>
+#endif
+#include "drivers/vulkan/godot_vulkan.h"
 
 static inline bool IsDuplicatePnext(VkStructureType input_value) {
     switch (input_value) {
