diff --git a/thirdparty/vulkan/vk_enum_string_helper.h b/thirdparty/vulkan/vk_enum_string_helper.h
index 9d2af46344..d61dbb1290 100644
--- a/thirdparty/vulkan/vk_enum_string_helper.h
+++ b/thirdparty/vulkan/vk_enum_string_helper.h
@@ -13,7 +13,11 @@
 #ifdef __cplusplus
 #include <string>
 #endif
-#include <vulkan/vulkan.h>
+#ifdef USE_VOLK
+    #include <volk.h>
+#else
+    #include <vulkan/vulkan.h>
+#endif
 static inline const char* string_VkResult(VkResult input_value) {
     switch (input_value) {
         case VK_SUCCESS:
