diff --git a/thirdparty/vulkan/vk_enum_string_helper.h b/thirdparty/vulkan/vk_enum_string_helper.h
index 062c3d9915..10df8e8115 100644
--- a/thirdparty/vulkan/vk_enum_string_helper.h
+++ b/thirdparty/vulkan/vk_enum_string_helper.h
@@ -28,8 +28,10 @@
 #pragma warning( disable : 4065 )
 #endif
 
+#ifdef __cplusplus
 #include <string>
-#include <vulkan/vulkan.h>
+#endif
+#include "drivers/vulkan/godot_vulkan.h"
 
 
 static inline const char* string_VkResult(VkResult input_value)
