diff --git a/thirdparty/vulkan/vk_enum_string_helper.h b/thirdparty/vulkan/vk_enum_string_helper.h
index 8c77dd3ee5..be968c091d 100644
--- a/thirdparty/vulkan/vk_enum_string_helper.h
+++ b/thirdparty/vulkan/vk_enum_string_helper.h
@@ -23,7 +23,11 @@
 // NOLINTBEGIN
 #pragma once
 #include <string>
-#include <vulkan/vulkan.h>
+#ifdef USE_VOLK
+    #include <volk.h>
+#else
+    #include <vulkan/vulkan.h>
+#endif
 
 static inline bool IsDuplicatePnext(VkStructureType input_value) {
     switch (input_value) {
