diff --git a/thirdparty/vulkan/vk_mem_alloc.h b/thirdparty/vulkan/vk_mem_alloc.h
index 8a42699e7f..510fa4d3ef 100644
--- a/thirdparty/vulkan/vk_mem_alloc.h
+++ b/thirdparty/vulkan/vk_mem_alloc.h
@@ -1771,7 +1771,11 @@ available through VmaAllocatorCreateInfo::pRecordSettings.
 #endif
 
 #ifndef VULKAN_H_
-    #include <vulkan/vulkan.h>
+    #ifdef USE_VOLK
+        #include <volk.h>
+    #else
+        #include <vulkan/vulkan.h>
+    #endif
 #endif
 
 #if VMA_RECORDING_ENABLED
