diff --git a/thirdparty/vulkan/vk_mem_alloc.h b/thirdparty/vulkan/vk_mem_alloc.h
index fa4298b23c..146895fd90 100644
--- a/thirdparty/vulkan/vk_mem_alloc.h
+++ b/thirdparty/vulkan/vk_mem_alloc.h
@@ -126,7 +126,7 @@ See documentation chapter: \ref statistics.
 extern "C" {
 #endif
 
-#include <vulkan/vulkan.h>
+#include "drivers/vulkan/godot_vulkan.h"
 
 #if !defined(VMA_VULKAN_VERSION)
     #if defined(VK_VERSION_1_3)
