diff -ruN original/Foundation/NSSharedPtr.hpp patched/Foundation/NSSharedPtr.hpp
--- original/Foundation/NSSharedPtr.hpp	2026-09-12 13:18:42
+++ patched/Foundation/NSSharedPtr.hpp	2026-09-12 13:18:42
@@ -22,6 +22,7 @@
 
 #include <cstddef>
 #include "NSDefines.hpp"
+#include "NSObject.hpp"
 
 namespace NS
 {
@@ -139,7 +140,7 @@
 _NS_INLINE NS::SharedPtr<_Class> RetainPtr(_Class* pObject)
 {
     NS::SharedPtr<_Class> ret;
-    ret.m_pObject = pObject->retain();
+    ret.m_pObject = Object::sendMessage<_Class*>(pObject, _NS_PRIVATE_SEL(retain));
     return ret;
 }
 
@@ -167,7 +168,7 @@
 template <class _Class>
 _NS_INLINE NS::SharedPtr<_Class>::~SharedPtr<_Class>() __attribute__((no_sanitize("undefined")))
 {
-    m_pObject->release();
+    Object::sendMessage<void>(m_pObject, _NS_PRIVATE_SEL(release));
 }
 
 template <class _Class>
@@ -178,14 +179,14 @@
 
 template <class _Class>
 _NS_INLINE NS::SharedPtr<_Class>::SharedPtr(const SharedPtr<_Class>& other) noexcept
-    : m_pObject(other.m_pObject->retain())
+    : m_pObject(Object::sendMessage<_Class*>(other.m_pObject, _NS_PRIVATE_SEL(retain)))
 {
 }
 
 template <class _Class>
 template <class _OtherClass>
 _NS_INLINE NS::SharedPtr<_Class>::SharedPtr(const SharedPtr<_OtherClass>& other, typename std::enable_if_t<std::is_convertible_v<_OtherClass *, _Class *>> *) noexcept
-    : m_pObject(reinterpret_cast<_Class*>(other.get()->retain()))
+    : m_pObject(Object::sendMessage<_Class*>(other.get(), _NS_PRIVATE_SEL(retain)))
 {
 }
 
@@ -225,7 +226,7 @@
 template <class _Class>
 _NS_INLINE void NS::SharedPtr<_Class>::reset() __attribute__((no_sanitize("undefined")))
 {
-    m_pObject->release();
+    Object::sendMessage<void>(m_pObject, _NS_PRIVATE_SEL(release));
     m_pObject = nullptr;
 }
 
@@ -240,9 +241,9 @@
 {
     _Class* pOldObject = m_pObject;
 
-    m_pObject = other.m_pObject->retain();
+    m_pObject = Object::sendMessage<_Class*>(other.m_pObject, _NS_PRIVATE_SEL(retain));
 
-    pOldObject->release();
+    Object::sendMessage<void>(pOldObject, _NS_PRIVATE_SEL(release));
 
     return *this;
 }
@@ -254,9 +255,9 @@
 {
     _Class* pOldObject = m_pObject;
 
-    m_pObject = reinterpret_cast<_Class*>(other.get()->retain());
+    m_pObject = Object::sendMessage<_Class*>(other.get(), _NS_PRIVATE_SEL(retain));
 
-    pOldObject->release();
+    Object::sendMessage<void>(pOldObject, _NS_PRIVATE_SEL(release));
 
     return *this;
 }
@@ -266,13 +267,13 @@
 {
     if (m_pObject != other.m_pObject)
     {
-        m_pObject->release();
+        Object::sendMessage<void>(m_pObject, _NS_PRIVATE_SEL(release));
         m_pObject = other.m_pObject;
     }
     else
     {
         m_pObject = other.m_pObject;
-        other.m_pObject->release();
+        Object::sendMessage<void>(other.m_pObject, _NS_PRIVATE_SEL(release));
     }
     other.m_pObject = nullptr;
     return *this;
@@ -285,7 +286,7 @@
 {
     if (m_pObject != other.get())
     {
-        m_pObject->release();
+        Object::sendMessage<void>(m_pObject, _NS_PRIVATE_SEL(release));
         m_pObject = reinterpret_cast<_Class*>(other.get());
         other.detach();
     }
