diff --git a/thirdparty/wslay/wslay/wslay.h b/thirdparty/wslay/wslay/wslay.h
index 7ab850111a..7d63bdb2ce 100644
--- a/thirdparty/wslay/wslay/wslay.h
+++ b/thirdparty/wslay/wslay/wslay.h
@@ -33,6 +33,11 @@ extern "C" {
 #include <stdlib.h>
 #include <sys/types.h>
 
+#if defined(_MSC_VER)
+#include <BaseTsd.h>
+typedef SSIZE_T ssize_t;
+#endif
+
 /*
  * wslay/wslayver.h is generated from wslay/wslayver.h.in by
  * configure. The projects which do not use autotools can set
