diff --git a/thirdparty/bullet/BulletSoftBody/btSparseSDF.h b/thirdparty/bullet/BulletSoftBody/btSparseSDF.h
index ae1288d9e6..243b80f8ae 100644
--- a/thirdparty/bullet/BulletSoftBody/btSparseSDF.h
+++ b/thirdparty/bullet/BulletSoftBody/btSparseSDF.h
@@ -233,9 +233,9 @@ struct btSparseSdf
 			//int sz = sizeof(Cell);
 			if (ncells > m_clampCells)
 			{
-				static int numResets = 0;
-				numResets++;
-				//				printf("numResets=%d\n",numResets);
+				//static int numResets = 0;
+				//numResets++;
+				//printf("numResets=%d\n",numResets);
 				Reset();
 			}
 
