diff --git a/thirdparty/libdivsufsort/include/config.h b/thirdparty/libdivsufsort/include/config.h
index 6e43b26a0e..acdf517547 100644
--- a/thirdparty/libdivsufsort/include/config.h
+++ b/thirdparty/libdivsufsort/include/config.h
@@ -40,7 +40,7 @@ extern "C" {
 #define HAVE_STDINT_H 1
 #define HAVE_STDLIB_H 1
 #define HAVE_STRING_H 1
-#define HAVE_STRINGS_H 1
+#define HAVE_STRINGS_H 0
 #define HAVE_MEMORY_H 1
 #define HAVE_SYS_TYPES_H 1
 
