diff --git a/thirdparty/re-spirv/re-spirv.cpp b/thirdparty/re-spirv/re-spirv.cpp
index fa613383b6..4dbf594bc3 100644
--- a/thirdparty/re-spirv/re-spirv.cpp
+++ b/thirdparty/re-spirv/re-spirv.cpp
@@ -2141,7 +2141,11 @@ namespace respv {
         std::vector<uint8_t> &optimizedData;
         Options options;
 
-        OptimizerContext() = delete;
+        OptimizerContext(const Shader &shader, std::vector<uint32_t> &instructionAdjacentListIndices, std::vector<uint32_t> &instructionInDegrees, std::vector<uint32_t> &instructionOutDegrees, std::vector<ListNode> &listNodes, std::vector<Resolution> &resolutions, std::vector<uint8_t> &optimizedData, Options options) :
+            shader(shader), instructionAdjacentListIndices(instructionAdjacentListIndices), instructionInDegrees(instructionInDegrees), instructionOutDegrees(instructionOutDegrees), listNodes(listNodes), resolutions(resolutions), optimizedData(optimizedData), options(options)
+        {
+                // Regular constructor.
+        }
     };
 
     static void optimizerEliminateInstruction(uint32_t pInstructionIndex, OptimizerContext &rContext) {
