diff --git a/thirdparty/re-spirv/re-spirv.cpp b/thirdparty/re-spirv/re-spirv.cpp
index 2815e40866b..1207b11cfee 100644
--- a/thirdparty/re-spirv/re-spirv.cpp
+++ b/thirdparty/re-spirv/re-spirv.cpp
@@ -2310,8 +2310,6 @@ namespace respv {
         std::vector<Resolution> &resolutions;
         std::vector<uint8_t> &optimizedData;
         Options options;
-
-        OptimizerContext() = delete;
     };
 
     static void optimizerEliminateInstruction(uint32_t pInstructionIndex, OptimizerContext &rContext) {
