diff --git a/thirdparty/jpeg-compressor/jpge.cpp b/thirdparty/jpeg-compressor/jpge.cpp
index 5a36c19653..bb0c54bbf0 100644
--- a/thirdparty/jpeg-compressor/jpge.cpp
+++ b/thirdparty/jpeg-compressor/jpge.cpp
@@ -30,6 +30,7 @@
 
 #include "jpge.h"
 
+#include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 
@@ -933,7 +934,6 @@ namespace jpge {
 	}
 
 	// Higher level wrappers/examples (optional).
-#include <stdio.h>
 
 	class cfile_stream : public output_stream
 	{
